ssl/t1_trce.c

```diff
@@ -1493,11 +1497,16 @@ void SSL_trace(int write_p, int version, int content_type,
     switch (content_type) {
     case SSL3_RT_HEADER:
         {
-            int hvers = msg[1] << 8 | msg[2];
+            int hvers;
+            if (msglen < 3)
+                return 0;
+            hvers = msg[1] << 8 | msg[2];
             BIO_puts(bio, write_p ? "Sent" : "Received");
             BIO_printf(bio, " Record\nHeader:\n  Version = %s (0x%x)\n",
                        ssl_trace_str(hvers, ssl_version_tbl), hvers);
             if (SSL_IS_DTLS(ssl)) {
+                if (msglen < 11)
+                    return 0;
```

Comment:
While we are at it, one can wonder if it would be appropriate to check even for overlap. I mean suggested modifications try to ensure that there is no buffer *overstepping* by asserting upper limit for message length. And question is whether or not it would be appropriate to assert even lower limit. Disregard SSL_IS_DTLS for a moment, and imagine that msglen is permitted[!] 3. In such case length field will overlap version field. Would it be appropriate to check for this? Well, it might happen that these conditions will always be favourable, even upper limit check, because this function might be called with data that is vetted already. But if you start playing the game, then maybe one should play it wholeheartedly...

Reply:
Interesting.
I guess that we want to avoid overlap (even if it is not an over read).
Anyone voting up for this solution ?